ntp-proto/src/algorithm/kalman/mod.rs

```diff
@@ -0,0 +1,358 @@
+use std::{collections::HashMap, fmt::Debug, hash::Hash};
+
+use tracing::{error, info, instrument};
+
+use crate::{
+    Measurement, NtpClock, NtpDuration, NtpLeapIndicator, NtpPacket, NtpTimestamp,
+    ObservablePeerTimedata, StateUpdate, SystemConfig, TimeSnapshot, TimeSyncController,
+};
+
+use self::{
+    combiner::combine,
+    config::AlgorithmConfig,
+    matrix::{Matrix, Vector},
+    peer::PeerState,
+};
+
+mod combiner;
+mod config;
+mod matrix;
+mod peer;
+mod select;
+
+fn sqr(x: f64) -> f64 {
+    x * x
+}
+
+#[derive(Debug, Clone)]
+struct PeerSnapshot<Index: Copy> {
+    index: Index,
+    state: Vector,
+    uncertainty: Matrix,
+    delay: f64,
+
+    peer_uncertainty: NtpDuration,
+    peer_delay: NtpDuration,
+    leap_indicator: NtpLeapIndicator,
+
+    last_update: NtpTimestamp,
+}
+
+impl<Index: Copy> PeerSnapshot<Index> {
+    fn offset(&self) -> f64 {
+        self.state.entry(0)
+    }
+
+    fn offset_uncertainty(&self) -> f64 {
+        self.uncertainty.entry(0, 0).sqrt()
+    }
+
+    fn observe(&self) -> ObservablePeerTimedata {
+        ObservablePeerTimedata {
+            offset: NtpDuration::from_seconds(self.offset()),
+            uncertainty: NtpDuration::from_seconds(self.offset_uncertainty()),
+            delay: NtpDuration::from_seconds(self.delay),
+            remote_delay: self.peer_delay,
+            remote_uncertainty: self.peer_uncertainty,
+            last_update: self.last_update,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct KalmanClockController<C: NtpClock, PeerID: Hash + Eq + Copy + Debug> {
+    peers: HashMap<PeerID, (PeerState, bool)>,
+    clock: C,
+    config: SystemConfig,
+    algo_config: AlgorithmConfig,
+    freq_offset: f64,
+    timedata: TimeSnapshot,
+    desired_freq: f64,
+    in_startup: bool,
+}
+
+impl<C: NtpClock, PeerID: Hash + Eq + Copy + Debug> KalmanClockController<C, PeerID> {
+    #[instrument(skip(self))]
+    fn update_peer(
+        &mut self,
+        id: PeerID,
+        measurement: Measurement,
+        packet: NtpPacket<'static>,
+    ) -> bool {
+        self.peers.get_mut(&id).map(|state| {
+            state
+                .0
+                .update(&self.config, &self.algo_config, measurement, packet)
+                & state.1
```

Comment:
this is `&` on bool, so I think `&&` is clearer

Reply:
Fixed